### ISSUE TITLE
Exception not handled when it involves ISP blocks

### ISSUE THREAD

**Issue by mrx23dot** [project-mk-ultra/torrent-tracker-scraper#31]
```
from torrent_tracker_scraper import scraper
scr = scraper.Scraper(infohashes=['1508da9bae10f29050a2894202bc78aab515799b'])
res = scr.scrape()
```

```
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "C:\Python36\lib\multiprocessing\pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "C:\Python36\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "C:\Python36\lib\site-packages\torrent_tracker_scraper\scraper.py", line 143, in scrape_tracker
    res = self.connection.sock.recv(8 + (12 * len(self.infohashes)))
OSError: [WinError 10040] A message sent on a datagram socket was larger than the internal message buffer or some other n
eive a datagram into was smaller than the datagram itself

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python36\lib\site-packages\torrent_tracker_scraper\scraper.py", line 186, in scrape
    results = list(filter(lambda result: result != [], results.get()))
  File "C:\Python36\lib\multiprocessing\pool.py", line 670, in get
    raise self._value
OSError: [WinError 10040] A message sent on a datagram socket was larger than the internal message buffer or some other n
eive a datagram into was smaller than the datagram itself
```
```
from torrent_tracker_scraper import scraper
scr = scraper.Scraper(infohashes=['1508da9bae10f29050a2894202bc78aab515799b'])
res = scr.scrape()
```
```
multiprocessing.pool.RemoteTraceback:
Traceback (most recent call last):
  File "C:\Python36\lib\multiprocessing\pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "C:\Python36\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "C:\Python36\lib\site-packages\torrent_tracker_scraper\scraper.py", line 115, in scrape_tracker
    _, transaction_id, connection_id = struct.unpack(">LLQ", res)
struct.error: unpack requires a buffer of 16 bytes

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python36\lib\site-packages\torrent_tracker_scraper\scraper.py", line 186, in scrape
    results = list(filter(lambda result: result != [], results.get()))
  File "C:\Python36\lib\multiprocessing\pool.py", line 670, in get
    raise self._value
struct.error: unpack requires a buffer of 16 bytes
```

Edit 1: formatting

**Comment by zawi99** [comment]
Related to https://github.com/project-mk-ultra/torrent-tracker-scraper/issues/30

**Comment by 49e94b8f256530dc0d41f740dfe8a4c1** [comment]
@mrx23dot 

We tried reproducing the issue but this code works

```
from torrent_tracker_scraper import scraper
scr = scraper.Scraper(infohashes=['1508da9bae10f29050a2894202bc78aab515799b'])
res = scr.scrape()

```

Could you update torrent_tracker_scraper by running `pip install torrent-tracker-scraper`  and trying it again?

**Comment by mrx23dot** [comment]
I think I know why it's happening, in UK the ISP blocks a lot of torrent/tracker sites.
The connection code simply needs a generic exception, in case anything bad happens.


Successfully installed torrent-tracker-scraper-3.1.2

ERROR:torrent_tracker_scraper.scraper:Socket timeout for tracker.opentrackr.org:1337: **timed out**
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "C:\Python36\lib\multiprocessing\pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "C:\Python36\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "C:\Python36\lib\site-packages\torrent_tracker_scraper\scraper.py", line 242, in scrape_tracker
    results = self._scrape_response(transaction_id, connection_id)
  File "C:\Python36\lib\site-packages\torrent_tracker_scraper\scraper.py", line 159, in _scrape_response
    res = self.connection.sock.recv(8 + (12 * len(self.good_infohashes)))
OSError: [WinError 10040] A message sent on a datagram socket was larger than the internal message buffer or some other network limit, or the buffer used to receive a datagram into was smaller than the datagram itself

**Comment by 49e94b8f256530dc0d41f740dfe8a4c1** [comment]
@mrx23dot,

We already handle network exceptions, but they are only thrown when there is a real network error

If it is true that your ISP is imposing restrictions on what their paying subscribers are allowed to access, then they are not blocking the trackers but just replaced the DNS addresses with their own servers and therefore redirecting  the request to said servers. This is how they show those "Piracy is a crime" scare websites. Perhaps you can confirm this by using `traceroute `and looking up where the route lands.
 
The traceback show that we indeed connect to the given tracker address, but since it is not a real tracker, the response does not conform to the torrent protocol hence the error. We see two errors because the code runs inside a thread. 

Could you confirm this speculation by going to http://opentrackr.org/ with your Internet browser please? 

You are right in that we did not account for this case. We therefore need to check responses before proceeding with the binary unpacking. Thank you for giving us more information. It will be very helpful in resolving this issue. 

I don't know the situation in the UK but you could try using a DNS server such as `1.1.1.1` or `8.8.8.8`?

Have a nice evening.

**Comment by mrx23dot** [comment]
To cover everything connection should look like:

```
try
  r = request.get(timeout=10)
  r.parse()
except:
 pass # anything can go wrong

```

DNS server is 8.8.8.8
I can connect to http://opentrackr.org/
But not to eg https://torrentz2.is/ (gives ERR_CONNECTION_CLOSED)

Thanks for the support!

**Comment by 49e94b8f256530dc0d41f740dfe8a4c1** [comment]
Replace 

```
try:
    res = self.connection.sock.recv(8 + (12 * len(self.good_infohashes)))
except socket.timeout as e:
    logger.error("Socket timeout for %s: %s", self.connection, e)
    return ["Socket timeout for %s: %s" % (self.connection, e)]
```

and 

` res = self.connection.sock.recv(16)`

With more robust exception handling by accounting for more exceptions, not just socket.timeout

**Comment by 49e94b8f256530dc0d41f740dfe8a4c1** [comment]
Related #46

**Comment by 49e94b8f256530dc0d41f740dfe8a4c1** [comment]
Fixed in #46
